runtime/major_gc.c

```diff
@@ -554,10 +563,136 @@ static void mark_ephe_aux (struct mark_stack *stk, intnat *work,
   }
 }
 
+
+#define Pb_size 256
+#define Pb_min 64
+#define Pb_mask (Pb_size - 1)
+#define Queue_prefetch_distance 4
+
+static uintnat rotate1(uintnat x)
+{
+  return (x << ((sizeof x)*8 - 1)) | (x >> 1);
+}
+
+CAMLnoinline static intnat do_some_marking(intnat work)
+{
+  uintnat pb_enqueued = 0, pb_dequeued = 0;
+  int darkened_anything = 0;
+  value pb[Pb_size];
+  uintnat min_pb = Pb_min;
+  struct mark_stack stk = *Caml_state->mark_stack;
+
+  uintnat young_start = (uintnat)Caml_state->young_start;
+  uintnat half_young_len = ((uintnat)Caml_state->young_end - (uintnat)Caml_state->young_start) >> 1;
+#define Is_block_and_not_young(v) \
+  (((intnat)rotate1((uintnat)v - young_start)) > (intnat)half_young_len)
+#ifdef NO_NAKED_POINTERS
+  #define Is_major_block(v) Is_block_and_not_young(v)
+  //#define Is_major_block(v) (Is_block(v) && !Is_young(v))
+#else
+#define Is_major_block(v) (Is_block_and_not_young(v) && Is_in_heap(v))
+#endif
+
+  while (1) {
+    value *scan, *obj_end, *scan_end;
+
+    if (pb_enqueued > pb_dequeued + min_pb) {
+      /* Dequeue from prefetch buffer */
+      value block = pb[(pb_dequeued++) & Pb_mask];
+      header_t hd = Hd_val(block);
+
+      /* FIXME: Forward_tag */
+      if (Tag_hd(hd) == Forward_tag) {
+
+      } else if (Tag_hd(hd) == Infix_tag) {
+        block -= Infix_offset_val(block);
+        hd = Hd_val(block);
+      }
+
+#ifdef NO_NAKED_POINTERS
+      /* See [caml_darken] for a description of this assertion. */
+      CAMLassert (Is_in_heap (block) || Is_black_hd (hd));
+#endif
+      CAMLassert(Is_white_hd(hd) || Is_black_hd(hd));
+      if (!Is_white_hd (hd)) {
+        /* Already black, nothing to do */
+        continue;
+      }
+      // FIXME work accounting
+      hd = Blackhd_hd (hd);
+      Hd_val (block) = hd;
+      darkened_anything = 1;
+      work--; // header word
+      if (Tag_hd (hd) >= No_scan_tag) {
+        /* Nothing to scan here */
+        work -= Wosize_hd (hd);
+        continue;
+      }
+      scan = Op_val(block);
+      obj_end = scan + Wosize_hd(hd);
+
+      if (Tag_hd (hd) == Closure_tag) {
+        uintnat env_offset = Start_env_closinfo(Closinfo_val(block));
+        work -= env_offset;
+        scan += env_offset;
+      }
+    } else if (work <= 0 || stk.count == 0) {
+      if (min_pb > 0) {
+        // drain pb before quitting
```

Comment:
If I understand what's going on, you're draining `pb`  by pushing everything back to the stack. Wouldn't it be faster to do it explicitly here with a small loop rather than go through the generic code below?

Reply:
It's slightly more complicated than that. If we reach this code because `work <= 0`, then you're right.

But if we reach this code because `stk.count == 0`, we have not yet reached the end. It's possible that there's some bottleneck in the heap layout, and the entire rest of the heap is accessible only through a half-dozen pointers, all of which happen to be in the prefetch buffer leaving the mark stack empty.

So, we should switch to a mode where we continue marking, but try to drain the prefetch buffer. The most likely outcome is that we do a couple more iterations before hitting this code again with `min_pb == 0` and quitting. However, it's also possible that we discover a huge amount more marking work by following the pointers in the prefetch buffer.

Reply:
Maybe the flow would be easier to follow if you separated the two cases:
```
if (work <= 0){
  // push everything back to stack and quit
}else if (pb_enqueued > pb.dequeued + min_pb || stk.count == 0 && pb_enqueued > bp_dequeued){
  // Dequeue from prefetch buffer
}else if (stk.count > 0){
  // take something from the stack
}else{
  // we're done
}
```
Just a suggestion. If you don't like it, I won't insist.

Reply:
I'm looking at this, but I don't currently see any easy way to make this change without duplicating lots of scanning logic for the `push everything back to stack and quit` path. This path isn't trivial: entries in the prefetch buffer are possibly-unmarked headers, while entries in the mark stack are intervals needing scanning. Pushing an entry from the prefetch buffer to the stack requires darkening, and dealing with No_scan_tag, Infix_tag and closure offsets.